File: src/org/spbrug/meetups/november2019/kotlin/_misc/_04_Interop.java
package org.spbrug.meetups.november2019.kotlin._misc;

import org.spbrug.meetups.november2019.kotlin.Customer;
import org.spbrug.meetups.november2019.kotlin.CustomerUtility;
import org.spbrug.meetups.november2019.kotlin.Status;

import static org.spbrug.meetups.november2019.kotlin.CustomerUtility.CopyrightYear;

public class _04_Interop {
    public static void main(String[] args) {
        Customer customer = new Customer(1, "Foo", Status.NEW);
        customer.changeStatus(Status.PENDING);

        System.out.println(CopyrightYear);

        CustomerUtility.extension(customer);

        try {
            customer.loadStatistics("");
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
        }
    }
}

File: src/org/spbrug/meetups/november2019/kotlin/_misc/_04_Interop.kt
@file:JvmName("CustomerUtility")
package org.spbrug.meetups.november2019.kotlin

const val CopyrightYear = 2019

data class Customer(val id: Int, var name: String, var status: Status = Status.NEW) {
    @JvmField
    val kind = "customer"

    @JvmOverloads
    fun changeStatus(newStatus: Status = Status.PENDING) {
        status = newStatus
    }

    @JvmName("preferential")
    fun makePreferred() {

    }

    @Throws(java.lang.IllegalArgumentException::class)
    fun loadStatistics(filename: String) {
        if (filename == "") {
            throw IllegalArgumentException("Filename can not be empty")
        }
    }
}

fun Customer.extension() = println("Extension called")

enum class Status {
    NEW,
    PENDING,
    PROCESSED
}

